File: proposed-projects/round-trip-experiment/reverse/20260909T184722Z_089a4294/20260910T101353Z_4b392f72/ported_implementation/src/utils/errors/input-parse-error.ts
import type { ValidInput } from '../../grammar-graph/types';
import { buildErrorPosition } from './build-error-position';
import { getInputAsString } from './get-input-as-string';

export const INPUT_PARSER_ERROR_HEADER_MESSAGE = 'Failed to parse input string:';

export class InputParseError extends Error {
  override name = 'InputParseError';
  #mostRecentInput: ValidInput;
  #pos: number;
  #previousInput: ValidInput;

  constructor(mostRecentInput: ValidInput, pos: number, previousInput: ValidInput = '') {
    const previous = getInputAsString(previousInput);
    super([
      INPUT_PARSER_ERROR_HEADER_MESSAGE,
      '',
      ...buildErrorPosition(
        `${previous}${getInputAsString(mostRecentInput)}`,
        pos + previous.length,
      ),
    ].join('\n'));
    this.#mostRecentInput = mostRecentInput;
    this.#pos = pos;
    this.#previousInput = previousInput;
  }

  get src(): string {
    return `${getInputAsString(this.#previousInput)}${getInputAsString(this.#mostRecentInput)}`;
  }

  get errorForMostRecentInput(): string {
    return [
      INPUT_PARSER_ERROR_HEADER_MESSAGE,
      '',
      ...buildErrorPosition(getInputAsString(this.#mostRecentInput), this.#pos),
    ].join('\n');
  }
}

File: proposed-projects/round-trip-experiment/reverse/20260909T123239Z_5982b634/20260910T070936Z_fb630842/ported_implementation/gbnf/utils/errors/input_parse_error.py
from __future__ import annotations

from .build_error_position import build_error_position
from .get_input_as_string import get_input_as_string
from .types import ValidInput

INPUT_PARSER_ERROR_HEADER_MESSAGE = 'Failed to parse input string:'


class InputParseError(Exception):
    def __init__(
        self,
        most_recent_input: ValidInput,
        pos: int,
        previous_input: ValidInput = '',
    ):
        previous = get_input_as_string(previous_input)
        super().__init__(
            '\n'.join(
                [
                    INPUT_PARSER_ERROR_HEADER_MESSAGE,
                    '',
                    *build_error_position(
                        f'{previous}{get_input_as_string(most_recent_input)}',
                        pos + len(previous),
                    ),
                ]
            )
        )
        self.name = 'InputParseError'
        self.most_recent_input = most_recent_input
        self.pos = pos
        self.previous_input = previous_input

    @property
    def src(self) -> str:
        return (
            f'{get_input_as_string(self.previous_input)}'
            f'{get_input_as_string(self.most_recent_input)}'
        )

    @property
    def error_for_most_recent_input(self) -> str:
        return '\n'.join(
            [
                INPUT_PARSER_ERROR_HEADER_MESSAGE,
                '',
                *build_error_position(
                    get_input_as_string(self.most_recent_input), self.pos
                ),
            ]
        )

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, InputParseError):
            return NotImplemented
        # The most recent input may be held as a string or as code points; compare
        # the rendered form so the two representations line up.
        return (
            self.pos == other.pos
            and self.src == other.src
            and str(self) == str(other)
        )

    def __hash__(self) -> int:
        return hash((self.pos, self.src))
